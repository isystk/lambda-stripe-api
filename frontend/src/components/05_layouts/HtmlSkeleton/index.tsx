import React, { Children, FC, useState } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import useAppRoot from '@/stores/useAppRoot'
import Title, { TitleProps } from './Title'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { APP_NAME, APP_DESCRIPTION } from '@/constants'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren & {
  showSideMenu?: boolean
}
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: TitleProps['children']
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  showSideMenu,
  main,
  title,
  description,
  children,
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* PCやモバイル（スマホ、タブレット）などのデバイスごとのコンテンツの表示領域 */}
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
      {/* Tailwind UI で採用されているカスタムフォント */}
      <link rel="stylesheet" href="https://rsms.me/inter/inter.css" />
    </Head>
    <Context.Provider value={main}>{children}</Context.Provider>
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  const main = useAppRoot()

  if (!main) return <></>

  children = Children.map(children, (child) =>
    isReactElement(child) && child.type === Title
      ? (title = `${child.props.children} | ${APP_NAME}`) && undefined
      : child
  )
  if (!title) title = APP_NAME

  const description = APP_DESCRIPTION
  return presenter({
    main,
    title,
    description,
    children,
    ...props,
  })
}

export const Context = React.createContext(null)

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
