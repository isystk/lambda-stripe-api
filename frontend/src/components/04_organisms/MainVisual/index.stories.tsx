import { Meta, Story } from '@storybook/react'
import React from 'react'
import MainVisual from './index'
import MainService from '@/services/main'
import { Context } from '@/components/05_layouts/HtmlSkeleton'

export default {
  title: '04_organisms/MainVisual',
  component: MainVisual,
} as Meta

const Template: Story = (props) => {
  const main = new MainService(() => ({}))

  return (
    <Context.Provider value={main}>
      <MainVisual />
    </Context.Provider>
  )
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
