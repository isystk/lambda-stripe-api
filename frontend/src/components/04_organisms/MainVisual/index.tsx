import React, { FC, useContext } from 'react'
import { ContainerProps, WithChildren } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'
import MainService from '@/services/main'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import LineButton from '@/components/01_atoms/LineButton'

/** MainVisualProps Props */
export type MainVisualProps = WithChildren
/** Presenter Props */
export type PresenterProps = MainVisualProps & {
  main
}

/** Presenter Component */
const MainVisualPresenter: FC<PresenterProps> = ({ main, ...props }) => (
  <>
    <section className={`${styles.mainVisual} h-screen`}>
      <div className="flex flex-wrap items-center h-full bg-main justify-center">
        <div className="flex w-full md:w-1/2 justify-center md:justify-end md:pr-16 pb-8">
          <img src="/images/iphone.png" className="w-48 md:w-72"></img>
        </div>
        <div className="flex w-full md:w-1/2 justify-center md:justify-start pb-8">
          <div className="px-8">
            <p className="text-accent text-xl md:text-3xl font-bold mb-6 ">
              リリース1ヶ月で100万人登録！
            </p>
            <h1 className="text-gray-700 text-2xl md:text-4xl font-bold mb-6 ">
              話題の○○が
              <br />
              LINEで使える！
            </h1>
            <LineButton link="#" label="友達に追加して質問してみる" />
          </div>
        </div>
      </div>
    </section>
  </>
)

/** Container Component */
const MainVisualContainer: React.FC<
  ContainerProps<MainVisualProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>

  return presenter({
    children,
    main,
    ...props,
  })
}

export default connect<MainVisualProps, PresenterProps>(
  'MainVisual',
  MainVisualPresenter,
  MainVisualContainer
)
