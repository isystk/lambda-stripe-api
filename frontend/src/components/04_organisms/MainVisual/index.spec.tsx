import React from 'react'
import renderer from 'react-test-renderer'
import MainVisual from './index'
import { Context } from '@/components/05_layouts/HtmlSkeleton'

import MainService from '@/services/main'

describe('MainVisual', () => {
  it('Match Snapshot', () => {
    const main = new MainService(() => ({}))

    const component = renderer.create(
      <Context.Provider value={main}>
        <MainVisual />
      </Context.Provider>
    )
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
