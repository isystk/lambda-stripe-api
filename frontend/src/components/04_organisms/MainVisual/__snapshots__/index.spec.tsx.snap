// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainVisual Match Snapshot 1`] = `
<section
  className="css-0 h-screen"
>
  <div
    className="flex flex-wrap items-center h-full bg-main justify-center"
  >
    <div
      className="flex w-full md:w-1/2 justify-center md:justify-end md:pr-16 pb-8"
    >
      <img
        className="w-48 md:w-72"
        src="/images/iphone.png"
      />
    </div>
    <div
      className="flex w-full md:w-1/2 justify-center md:justify-start pb-8"
    >
      <div
        className="px-8"
      >
        <p
          className="text-accent text-xl md:text-3xl font-bold mb-6 "
        >
          リリース1ヶ月で100万人登録！
        </p>
        <h1
          className="text-gray-700 text-2xl md:text-4xl font-bold mb-6 "
        >
          話題の○○が
          <br />
          LINEで使える！
        </h1>
        <a
          className="css-0 p-2 pl-4 pr-4 flex items-center justify-center bg-accent rounded-full"
          href="#"
        >
          <img
            height="50px"
            src="/images/icon-line.png"
            width="50px"
          />
          <span
            className="text-white font-bold"
          >
            友達に追加して質問してみる
          </span>
        </a>
      </div>
    </div>
  </div>
</section>
`;
